### PR TITLE
add december blog post

### PR DESCRIPTION
**Pull request by locks**
My blog post for December. This is intended as a series, hence the mentioned to further blog posts in the copy.
~I am unsure if the section on DOM events should be included in this blog post or its own separate blog post, that is why only the header is included at the moment. Please leave your thoughts!~
I have removed the section on DOM events, it will be covered in a follow-up post.

Planned follow ups:
- `action` -> `on` + `fn`
- element modifiers (alluded to in `action`, `on`, `ember-modifiers`, `ember-render-modifiers`)
- Glimmer components/template only components (API simplification, element modifiers instead of component lifecycle hooks)

### CODE REVIEW

File: _posts/2019-12-23-clarity-in-templates.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+
+In this blog post we will be discussing how recent Ember.js modernization efforts on the path to Ember Octane have brought features that help with this exact problem.
+
+<!--break-->
+
+![invest time open source illustration](/assets/images/posts/2019-11-11-why-companies-invest-in-oss/oos-illustration.png#full)
+
+## Ember Octane and editions
+
+As a framework that values stability and providing its users solid upgrade paths, Ember has accrued its fair share of idiosyncrasies. One of the more demonstrable consequences of this is the so called pit of incoherence, alluded to in the opening keynote of EmberConf 2019 [[1](https://www.youtube.com/watch?v=zYwdBcmz6VI): APIs are introduced that move the framework to a new mental model, but that mental model isn't complete yet. This leaves users unsure of which APIs to use when, and how the interop might work.
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: I think the link is malformed - at least it shows up wrong on the rendered page:

Bildschirmfoto 2019-12-17 um 14 54 15

I guess it'd be nicer for the reader anyway when the phrase _"opening keynote of EmberConf 2019"_ was linked anyway?

File: _posts/2019-12-23-clarity-in-templates.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+
+### Disambiguating named arguments
+
+Next we will turn our attention to dynamic values that were passed into the component when it was invoked. These are called named arguments. This time we will look at an invocation of our `blog/post` component:
+
+```hbs
+{{! app/templates/blog.hbs }}
+{{blog/post post=post onReportComment=(action 'onReportComment')}}
+```
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: I'm not sure I understand why this template is introduced when you're introducing the `@` in the previously used template then?

File: _posts/2019-12-23-clarity-in-templates.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+<button {{action 'expandComments'}}>
+
+{{#if this.commentsExpanded}}
+  {{#each post.comments as |commentObject|}}
+    {{comment comment=commentObject onReport=@onReportComments}}
+  {{/each}}
+{{/if}}
+```
+
+We have improved our template a further step. Now we can tell properties, named arguments, and local variables or helpers/components apart.
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: I think _"helpers/components"_ are not actually covered at this point?

File: _posts/2019-12-23-clarity-in-templates.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+
+If you wish to use named arguments in your application, you can use ember-named-arguments-polyfill [[x](https://github.com/rwjblue/ember-named-arguments-polyfill)] for Ember.js versions older than 3.1.
+
+Note that `@model` was introduced by RFC #523 [[x](https://emberjs.github.io/rfcs/0523-model-argument-for-route-templates.html)] so that you can refer directly to the model that was passed to a route template.
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: I'd expand this a bit to make it clearer. The point is that the mechanism for separating named arguments obviously only works for components as those are the only templates in Ember that can receive arguments at all and with RFC #523 that mechanism is made available for route templates as well, effectively treating the route's model like a named argument for the template. Of course that's clear to us but I'm not sure this analogy is obvious to all Ember developers.

File: _posts/2019-12-23-clarity-in-templates.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+### Disambiguating components
+
+After the last change, it would be nice to have a way to disambiguate component invocations from other kinds of dynamic interpolations. Fortunately, a new syntax was introduced called angle bracket invocation syntax [[x](https://emberjs.github.io/rfcs/0311-angle-bracket-invocation.html)] that allows us to do just that. It also allows us to distinguish between named arguments and HTML attributes when invoking components, but that will be covered in a future post.
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: > called angle bracket invocation syntax [[x](https://emberjs.github.io/rfcs/0311-angle-bracket-invocation.html)]

I'm generally not a fan of this linking style but tbh I think when using _"x"_ vs. numbers it becomes completely pointless?

Comment (marcoow) [reply _posts/2019-12-23-clarity-in-templates.md]: > It also allows us to distinguish between named arguments and HTML attributes when invoking components, but that will be covered in a future post.

You're introducing the concept of named arguments in the previous paragraph though? I understand you're referring to applying HTML attributes in the component here but this reads a bit strange as you think "he just explained named arguments?".

File: _posts/2019-12-23-clarity-in-templates.md [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+To use this feature in older versions of Ember and its dependencies, you can use `ember-angle-bracket-invocation-polyfilll` [[x](https://github.com/rwjblue/ember-angle-bracket-invocation-polyfill)].
+
+With just a couple of tweaks, we now have much more clarity when reading a template, and the need to consult additional files is lessened:
+- `{{this.title}}` is a property that comes from the JavaScript file of the component;
+- `<PostBody />` is a component invocation;
+- `{{@post}}` is a named argument that is passed to the component when invoked;
+- `{{this.commentsExpanded}}` is also a property;
+- `if` and `each` are templating constructs
+- `commentObject` is a block argument that is available inside the `each` block scope
+- `Comment` is a component invocation
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]:
```suggestion
- `<Comment />` is a component invocation
```

File: _posts/2019-12-23-clarity-in-templates.md [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+## Handling DOM events
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: I assume this section should be removed?

**Review by marcoow — COMMENTED** [review]
Nice post! I just left a few comments but in general this reads very well. Here's some other notes:

- you need to add your photo:

Bildschirmfoto 2019-12-17 um 15 09 48

- the post ends a bit abruptly, it might be good to have a closing sentence of some sort?
- this will need an illustration and an `og:image`

File: _posts/2019-12-23-clarity-in-templates.md (hunk 7 above)
Comment (locks) [reply _posts/2019-12-23-clarity-in-templates.md]: Confirm, replacing with an ending.

File: _posts/2019-12-23-clarity-in-templates.md (hunk 2 above)
Comment (locks) [reply _posts/2019-12-23-clarity-in-templates.md]: The idea was the same as introducing the class file in the properties section, to show that you need to consult a different file to understand what might be passed in, adding a bit more contex to `post` and `onReportComments`.

File: _posts/2019-12-23-clarity-in-templates.md (hunk 3 above)
Comment (locks) [reply _posts/2019-12-23-clarity-in-templates.md]: I mentioned them in the intro:

> Looking only at the template, can we be certain where `title` comes from? Whether `post-body` and `comment` are helpers or components? We can generously assume that `comment` is a component as it is receiving an action, but when scanning the file it is indistinguishable from `if`, `each`, and other templating constructs.

Is there something I can clarify?

File: _posts/2019-12-23-clarity-in-templates.md (hunk 5 above)
Comment (locks) [reply _posts/2019-12-23-clarity-in-templates.md]: I've updated the link style.

File: _posts/2019-12-23-clarity-in-templates.md (hunk 2 above)
Comment (marcoow) [reply _posts/2019-12-23-clarity-in-templates.md]: Ah, so you're saying the above template **is** the `blog/post` component and here you're showing the invocation? That makes sense but I don't think it's clear necessarily when reading through the text - might be worth adding a sentence here explaining that.

File: _posts/2019-12-23-clarity-in-templates.md (hunk 3 above)
Comment (marcoow) [reply _posts/2019-12-23-clarity-in-templates.md]: Yes, what I mean is we can only tell helpers/component apart from local variables with angle bracket invocation which hasn't been introduced at this point.

File: _posts/2019-12-23-clarity-in-templates.md [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+Next we will turn our attention to named arguments. We give this name to values that we pass into components when we use them. Since components do not have access to the context where they are called in, this is the primary way to pass information into them.
+
+To see what we are passing into our `blog/post` component, we need to see how it is used in our application. To do this, we will check a route template, namely the template for the `blog` route, where the component is used:
+
+```hbs
+{{! app/templates/blog.hbs }}
+{{blog/post post=post onReportComment=(action 'onReportComment')}}
+```
+
+We can see that when calling the component, we are passing in `post` and `onReportComment`. The syntax to mark them as named argument is a `@` prefix, so let us update our template:
```
Comment (marcoow) [inline _posts/2019-12-23-clarity-in-templates.md]: Will this actually work with the invocation shown above that does **not** use the `@` prefix?

**Review by marcoow — COMMENTED** [review]
Nice - made just one more comment about something I'm not sure about.

File: _posts/2019-12-23-clarity-in-templates.md (hunk 8 above)
Comment (locks) [reply _posts/2019-12-23-clarity-in-templates.md]: I believe so. In `@ember/component` passed in arguments are reflected as properties so  `this.onReportComment`, `onReportComment` and `@onReportComment` are roughly equivalent. With `@onReportComment` being directly the value passed in.

Comment (marcoow) [reply _posts/2019-12-23-clarity-in-templates.md]: Yep, this works indeed - just checked: https://ember-twiddle.com/a961c72d6959a5d36fe91e65f8d3323a?openFiles=templates.components.my-component.hbs%2Ctemplates.components.my-component.hbs

**Review by marcoow — APPROVED** [review]

**Comment by marcoow** [comment]
@locks: can you add your photo, the images that @ghislaineguerin created and update the date to tomorrow so we can get this out?

**Review by marcoow — CHANGES_REQUESTED** [review]
@locks can you make your profile image smaller? Smaller dimensions should be sufficient. Also can you run images through https://imageoptim.com/mac and fix CI (I think you're just missing `yarn format`)?

**Comment by locks** [comment]
@marcoow I've shrunken my profile image to 200x200 like the others and ran the png through imageoptim. `yarn format` is also applied.

**Review by marcoow — APPROVED** [review]

**Pull request merged**